source/agora/common/Types.d

```diff
@@ -92,6 +92,58 @@ public struct Address
     public alias inner this;
 }
 
+/// A simple wrapper struct to string-represent IPv4 without allocating
+public struct IPv4
+{
+    import std.format;
+
+    public uint value;
+
+    /// Returns: An IPv4 in network byte order
+    public static IPv4 fromHost (uint v) @safe pure nothrow @nogc
+    {
+        import core.sys.posix.arpa.inet : htonl;
+        return IPv4(htonl(v));
+    }
+
+    /// Provides a string representation of this IP
+    public void toString (scope void delegate(in char[]) @safe sink)
+        const scope @safe
+    {
+        formattedWrite!"%s.%s.%s.%s"(sink,
+            this.value >> 24 & 0xFF, this.value >> 16 & 0xFF,
+            this.value >>  8 & 0xFF, this.value >>  0 & 0xFF);
+    }
+
+    public string toString () const scope @safe
+    {
+        char["255.255.255.255".length] buffer;
+        size_t offset;
+        scope sink = (in char[] data) {
+            assert(offset + data.length <= buffer.length);
+            buffer[offset .. offset + data.length] = data[];
+            offset += data.length;
+        };
+        this.toString(sink);
+        return buffer[0 .. offset].idup;
+    }
+}
+
+///
+unittest
+{
+    IPv4 all = IPv4(0);
```

Comment:
would it be better to force use of `fromHost` or have a ctor?

Reply:
I was wondering this. Having a disabled ctor causes all sort of problem so I went with the simplest approach for the time being.